docs/code-guidelines.md

```diff
@@ -0,0 +1,308 @@
+2. It's hard to control/customize the outer, layout parts of the UI from the specific page view (example: breadcrumbs). You have to rely on `useEffect` and `React.Context`.
+
+Instead of doing that, we render most of the app UI from page components. A typical route/page pair looks like that:
+
+```jsx
+const Page = () => {
+  return (
+    <Layout>
+      <Container>
+        <View />
+      </Container>
+    </Layout>
+  );
+};
+
+<Route path="/page" element={<Page />} />;
+```
+
+One obvious downside of it should be that the whole Layout gets remounted when the Page changes even if the next Page renders the same type of Layout.
+However, we work around that limitation by "throwing" layout to the top of the tree, where it's rendered by the Layout Holder. For that technique to work, the following must be observed:
+
+1. There should be a `LayoutHolder` component rendered somewhere high in the tree, ideally above all routes. Please see `createLayoutHolder` in `src/core/ui/layout/LayoutHolder/LayoutHolder.tsx`
+2. The Layout rendered from a Page should be wrapped in a HOC `createLayout` (returned by `createLayoutHolder`).
+3. All Page components should return some type of Layout produced by `createLayout`.
+
+That ensures that as long as the type of the returned Layout stays the same, only its props get updated.
+
+> :information_source: For use within an index Route, use `createLayoutHolderWithOutlet` helper, which adapts `createLayoutHolder` to `react-router`.
+
+> :warning: One important side effect of the described architecture is that the hooks placed in the Page component body would not inherit this special behavior of Layout.
+> They will keep the normal behavior of React hooks and will be reinstantiated on each route/page change.
+> For efficient data loading it's key to keep all Apollo-related hooks in a `Container` and not place them directly into the `Page.`
+
+The name of a Container or a View for the given page should either
+- inherit the page name if its intended usage is exclusive to the Page it's built for;
+- have broader naming if they are or can be used in other pages or for other entities.
+
+Pseudocode example:
+```jsx
+const SpaceDashboardPage = () => {
+  return (
+    <JourneyLayout journeyTypeName="space"> // broader naming
+      <SpaceDashboardContainer> // specific to the entity/page
+        <JourneyDashboardView journeyTypeName="space" /> // generic entity yet specific to the page
+      </SpaceDashboardContainer>
+    </JourneyLayout>
+  );
+};
+```
+
+## Component Library and Avoidance of code repetition
+
+Any component that represents a self-sufficient UI building block and/or contains CSS tweaks, should be considered a UI Library Component and put into `src/core/ui`.
+Domain-specific components should normally not include low-level markup (such as rendering `<div>` or `<Box>`) or CSS with the exception of one-time customization/an exclusive override for a specific page.
+
+## To be done:
```

Comment:
I would like to have some kind of summary of all of this to take a look during CodeReviews, like a list of checks or something like that.
I guess somebody else can do it, probably myself :D